File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    

    def letterCombinations(self, digits: str) -> List[str]:
        keys={"2":"abc", "3":"def","4":"ghi","5":"jkl","6":"mno","7":"pqrs","8":"tuv","9":"wxyz"}
        res=[]
        combinations=""
        if not digits:
            return []
        def recursion(i,digits,combinations):
            if i==len(digits):
                res.append(combinations[:])
                return
            for letter in keys[digits[i]]:
                recursion(i+1,digits,combinations+letter)
        recursion(0,digits,combinations)
        return res

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void solve(string &digits, vector<string> &choice, vector<string> &result, string &ans, int index){

        if(ans.length() == digits.size()){
            result.push_back(ans);
            return;
        }

        int idx = digits[index] - '0';
        string temp = choice[idx-2];

        for(int i=0;i<temp.length();i++){
            
            ans.push_back(temp[i]);
            solve(digits, choice, result, ans, index+1);
            ans.pop_back();
        }

    }

    vector<string> letterCombinations(string digits) {

        if(digits.size() == 0)
            return {};

        vector<string> choice = {"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        vector<string> result;
        string ans = "";

        solve(digits, choice, result, ans, 0);

        return result;
    }
};